hermes_core/util/cdf.py

```diff
@@ -0,0 +1,612 @@
+        """
+        Function to Derive Attributes based on other attribues in the CDF
+        writer's internal dict state.
+        """
+        # Cache for easier access
+        method = attr_schema["derived"]
+
+        # SWITCH on the Derivation Method
+        if method == "get_generation_date":
+            return self._get_generation_date()
+        elif method == "get_start_time":
+            return self._get_start_time()
+        elif method == "get_data_type":
+            return self._get_data_type()
+        elif method == "get_logical_file_id":
+            return self._get_logical_file_id()
+        elif method == "get_logical_source":
+            return self._get_logical_source()
+        elif method == "get_logical_source_description":
+            return self._get_logical_source_description()
+        else:
+            raise ValueError(f"Derivation Method ({method}) Not Recognized")
+
+    def verify_global_attr_schema(self):
+        """
+        Function to ensure all required attributes in the schema are present
+        in the `self.global_attr` member.
+
+        Raises `ValueError` if there are required attributes in the schema that
+        are not in the global attributes.
+        """
+        # Loop for each attribute in the schema
+        for attr_name, attr_schema in self.schema.items():
+            # If it is a required attribute and not present
+            if attr_schema["required"] and (attr_name not in self.global_attrs):
+                raise ValueError(
+                    (
+                        f"Required attribute ({attr_name}) not present in global attributes.",
+                        "Make sure to add all required attributes to the gloabl attributes",
+                        "before genereating a CDF file.",
+                    )
+                )
+
+    def _convert_global_attributes_to_cdf(self):
+        # Loop though Global Attributes in target_dict
+        for attr_name, attr_value in self.global_attrs.items():
+            # Make sure the Value is not None
+            # We cannot add None Values to the CDF Global Attrs
+            if not attr_value:
+                warn_user((f"Cannot Add gAttr: {attr_name}. Value was {str(attr_value)}. "))
+                # Set to a intentionally invalid value
+                attr_value = f"{attr_name}-Not-Provided"
+            # Add the Attribute to the CDF File
+            self.target_cdf_file.attrs[attr_name] = attr_value
+
+    def _convert_variable_attributes_to_cdf(self):
+        # Loop through Variable Attributes in target_dict
+        for var_name, var_data, var_attrs in self.__iter__():
+            # Add the Variable to the CDF File
+            self.target_cdf_file[var_name] = var_data
+            # Add the Variable Attributes
+            for var_attr_name, var_attr_val in var_attrs.items():
+                self.target_cdf_file[var_name].attrs[var_attr_name] = var_attr_val
+
+    def save_cdf(self):
```

Comment:
Why do we need a `to_cdf` and a `save_cdf` function? Also i would recommend calling it `write_cdf`.

Reply:
Same as mentioned above: The goal here for splitting the to_cdf and save_cdf functions is to give users the chance to manipulate the pycdf data structure directly before closing the file.

Reply:
How we call the `to_cdf` function just `cdf` and could it be written to a temporary file? Then the write command would move or copy the temporary file to a real directory. What do you think?

Reply:
If we write to a temp file in the `to_cdf()` or just `cdf()` function then we will have to make sure to delete the temp file before calling the `write_cdf()` function, or else we might run into a name collision error. If we gave the temp file a different name, so we don’t run into the name collision error, then the `spacepy.ISTP` validation will fail which compares the filename against the `Logical_file_id` global metadata.  I don’t really see much benefit of having the first function create a temp file if `write_cdf()` is just going to delete the temp file and rewrite all the data to a new file. We may also end up getting out-of-sync here if someone creates a CDF using `to_cdf()`, modifies the `spacepy.CDF` directly, and then tries to `write_cdf()` . We would have to be careful in this scenario that we make sure the internal `TimeSeries` object is updated, along with the `spacepy.CDF`, before deleting the temp file otherwise those changes will be lost. What you’re suggesting is definitely possible, just could get a little ugly. 
I vote we either keep the current design, or we don’t allow modification if the `spacepy.CDF` data. Open to other ideas though.

Reply:
Having walked through the code with Andrew on a call, I think the separate modules of to_cdf, validate_cdf, and save_cdf make a lot of sense.